New function LISTAGG

ChudaykinAlex opened the pull request:
1) Purpose
The current implementation has an aggregate function LIST which concatenates multiple row fields into a blob. The SQL standard has a similar function called LISTAGG. The major difference is that it also supports the ordered concatenation.

2) Syntax and rules
```
<listagg set function> ::=
LISTAGG <left paren> [ <set quantifier> ] <character value expression> <comma> <listagg separator> [ <listagg overflow clause> ] <right paren> <within group specification>

<listagg separator> ::= 
<character string literal>

<listagg overflow clause> ::=
ON OVERFLOW <overflow behavior>

<overflow behavior> ::=
ERROR | TRUNCATE [ <listagg truncation filler> ] <listagg count indication>

<listagg truncation filler> ::=
<character string literal>

<listagg count indication> ::=
WITH COUNT | WITHOUT COUNT

<within group specification> ::=
WITHIN GROUP <left paren> ORDER BY <sort specification list> <right paren>
```
The legacy LIST syntax is preserved for backward compatibility, LISTAGG is added to cover the standard features.

There is a ```<listagg overflow clause>``` rule in the standard, which is intended to output an error when the output value overflows. Since the LIST function always returns a BLOB, it was decided that this rule would be meaningless. It was not implemented and silently ignored if specified.

If DISTINCT is specified for LISTAGG, then ORDER BY ```<sort specification list>``` must fully match ```<character value expression>```

If DISTINCT is specified, the presence of WITHIN GROUP must obey the restriction and will not affect the subsequent code execution.

3) Examples:
```
CREATE TABLE TEST_T
	(COL1 INT, COL2 VARCHAR(2), COL3 VARCHAR(2), COL4 VARCHAR(2), COL5 BOOLEAN, COL6 VARCHAR(2)
	CHARACTER SET WIN1251);
COMMIT;
INSERT INTO TEST_T values(1, 'A', 'A', 'J', false, 'П');
INSERT INTO TEST_T values(2, 'B', 'B', 'I', false, 'Д');
INSERT INTO TEST_T values(3, 'C', 'A', 'L', true,  'Ж');
INSERT INTO TEST_T values(4, 'D', 'B', 'K', true,  'Й');
COMMIT;

SELECT LISTAGG (ALL COL4, ':') AS FROM TEST_T;
=======
J:I:L:K

SELECT LISTAGG (DISTINCT COL4, ':') FROM TEST_T;
========
I:J:K:L

SELECT LISTAGG (DISTINCT COL3, ':') FROM TEST_T;
====
A:B

SELECT LISTAGG (DISTINCT COL3, ':') WITHIN GROUP (ORDER BY COL2) FROM TEST_T;
====
A:B

SELECT LISTAGG (DISTINCT COL3, ':') WITHIN GROUP (ORDER BY COL2 DESCENDING) FROM TEST_T;
====
A:B

SELECT LISTAGG (COL2, ':') WITHIN GROUP (ORDER BY COL2 DESCENDING) FROM TEST_T;
=======
D:C:B:A

SELECT LISTAGG (COL4, ':') WITHIN GROUP (ORDER BY COL3 DESC) FROM TEST_T;
=======
I:K:J:L

SELECT LISTAGG (COL3, ':') WITHIN GROUP (ORDER BY COL5 ASCENDING) FROM TEST_T;
=======
A:B:A:B

SELECT LISTAGG (COL4, ':') WITHIN GROUP (ORDER BY COL3 ASC) FROM TEST_T;
=======
J:L:I:K

SELECT LISTAGG (ALL COL2) WITHIN GROUP (ORDER BY COL4) FROM TEST_T;
=======
B,A,D,C

SELECT LISTAGG (COL2, ':') WITHIN GROUP (ORDER BY COL3 DESC, COL4 ASC) FROM TEST_T;
=======
B:D:A:C

SELECT LISTAGG (COL2, ':') WITHIN GROUP (ORDER BY COL3 DESC, COL4 DESC) FROM TEST_T;
=======
D:B:C:A

SELECT LISTAGG (COL2, ':') WITHIN GROUP (ORDER BY COL3 ASC, COL4 DESC) FROM TEST_T;
=======
C:A:D:B

SELECT LISTAGG (ALL COL6, ':')FROM TEST_T;
=======
П:Д:Ж:Й

SELECT LISTAGG (ALL COL6, ':') WITHIN GROUP (ORDER BY COL2 DESC) FROM TEST_T;
=======
Й:Ж:Д:П

SELECT LISTAGG (ALL COL2, ':') WITHIN GROUP (ORDER BY COL6) FROM TEST_T;
=======
B:C:D:A

INSERT INTO TEST_T values(5, 'E', NULL, NULL, NULL, NULL);
INSERT INTO TEST_T values(6, 'F', 'C', 'N', true, 'К');

SELECT LISTAGG (ALL COL2, ':') WITHIN GROUP (ORDER BY COL3) FROM TEST_T;
===========
E:A:C:B:D:F

SELECT LISTAGG (ALL COL2, ':') WITHIN GROUP (ORDER BY COL3 NULLS LAST) FROM TEST_T;
===========
A:C:B:D:F:E

SELECT LISTAGG (ALL COL2, ':') WITHIN GROUP (ORDER BY COL6 NULLS FIRST) FROM TEST_T;
===========
E:B:C:D:F:A

SELECT LISTAGG (DISTINCT COL3, ':') WITHIN GROUP (ORDER BY COL2) FROM TEST_T;
========
Statement failed, SQLSTATE = 42000
SQL error code = -104
-Invalid command
-Sort-key of the ORDER BY specification must match the argument list
```

ChudaykinAlex committed:
Adding an implementation of the new LISTAGG function

ChudaykinAlex committed:
Merge branch 'master' into work/listagg

ChudaykinAlex committed:
Add README

sim1984 commented:
This PR implements one of the aggregate functions (LISTAGG), which depends on the order of the input stream records. The full list can be seen here: https://github.com/FirebirdSQL/firebird/issues/7632

dyemanov commented on src/dsql/AggNodes.cpp:
```diff
@@ -307,7 +308,7 @@ bool AggNode::dsqlMatch(DsqlCompilerScratch* dsqlScratch, const ExprNode* other,
 	// per function.
 	return aggInfo.blr == o->aggInfo.blr && aggInfo.name == o->aggInfo.name &&
-		distinct == o->distinct && dialect1 == o->dialect1;
+		distinct == o->distinct && dialect1 == o->dialect1 && sort == o->sort;;
```
@ChudaykinAlex IMO, the sort node should be added to the list by `ListAggNode::getChildren() `-- this way you may remove the sort comparison in `dsqlMatch()` and also remove `doPass2(sort)` in `pass2()`, as the sort node will be processed by the inherited methods automagically.

dyemanov commented on src/dsql/AggNodes.cpp:
```diff
@@ -899,6 +944,26 @@ void ListAggNode::make(DsqlCompilerScratch* dsqlScratch, dsc* desc)
+	if (sort && distinct)
+	{
+		ValueExprNode* const sortNode = *sort->expressions.begin();
+		if (!arg->sameAs(sortNode, false) || sort->expressions.getCount() > 1)
```
Good point. Currently, we sort only once and this is a good bonus. If we follow your suggestion and allow slightly different expressions, then we should either sort twice or ignore the user-specified ordering after `DISTINCT`. BTW, in this PR it also seems to be ignored -- `LISTAGG(DISTINCT COL) WITHIN GROUP (ORDER BY COL DESC)` would produce ASC-ordered output. But I suppose it should work the same way as for plain `SELECT DISTINCT(COL) FROM T ORDER BY COL DESC`, i.e. respect the `ORDER BY ` ordering and optimize the sorts (merge two sorts into one) only if they fully (expressions / directions / NULLs placement) match each other.

dyemanov replied:
Ah, sorry, my bad. Surely, direction for the combined sort should be taken from `ORDER BY` -- like we already do in `Optimizer::checkSorts()`.

asfernandes replied:
> Nulls placement is not important, I agree, as NULLs are skipped by all aggregate functions.

With my suggestion to use same existing rules, the DISTINCT LISTAGG expression may be something like `COALESCE(field, 'z') with a `ORDER BY Z`.

dyemanov replied:
Given that we support queries like:

`select distinct emp_no from employee order by last_name`

I'm not sure we need to introduce the `GROUP BY` rules inside `LISTAGG (DISTINCT ...) ... WITHIN GROUP`. I suppose `DISTINCT + ORDER BY` should work the same way as either a standalone construct or within the `LISTAGG` group, i.e. without any restrictions. That said, I see no practical point in the sort behind the `ORDER BY <other expression>` clause because its result is not going to be user-visible anyway. It basically means that if `DISTINCT` is present, we may simply ignore the `ORDER BY` clause -- except when `ORDER BY` mentions the aggregated expression in the very beginning of its `<sort specification list>` -- then its sort direction should be taken into account inside `DISTINCT`. 

Do I miss anything?

ChudaykinAlex committed:
Merge branch 'master' into work/listagg

dyemanov commented on doc/sql.extensions/README.listagg:
```diff
@@ -0,0 +1,144 @@
+INSERT INTO TEST_T values(4, 'D', 'B', 'K', true,  'Й');
+COMMIT;
+
+SELECT LISTAGG (ALL COL4, ':') AS FROM TEST_T;
```
The SQL specification declares `<within group specification>` as mandatory. However, IMHO it's quite restrictive and neither Oracle nor DB2 follows that rule, they have it optional. Given that `LIST` and `LISTAGG` share the same syntax in this PR, we've also made `<within group specification>` optional. So the easiest solution is to fix the README ;-)

Or we may go the standard way and separate the legacy `LIST` (leave it with the current grammar, without ordering) from `LISTAGG` (which is strictly standard-compliant) at the parser level. But IMHO it would be annoying for users to select either of them depending on whether you need ordering or not. So personally I'd keep everything "as is" and just fix the docs.

Other opinions?

ChudaykinAlex replied:
Strange. I need to check it out

dyemanov replied:
If we remove `ON OVERFLOW` from the docs, then I believe we should remove it from the parser too.

ChudaykinAlex committed:
Code formatting errors have been fixed.
Adjustments have been made to the README.
Fixed a bug with the <listagg overflow clause> behavior, now it is silently ignored.
The dsqlMatch function has been redesigned.

Redesigned behavior with DISTINCT. Multiple elements are now allowed in the ORDER BY.
I also added influences on the sorting direction.

ChudaykinAlex committed:
dyemanov omments have been corrected.

ChudaykinAlex committed:
Merge branch 'master' into work/listagg
